### PR TITLE
feat: Add faiss

### PR DESCRIPTION
**Pull request by Pringled**

### CODE REVIEW

File: tests/conftest.py [hunk 1]
```diff
@@ -8,32 +8,58 @@
 
 random_gen = np.random.default_rng(42)
 
+FAISS_INDEX_TYPES = ["flat", "ivf", "hnsw", "lsh", "scalar", "pq", "ivf_scalar", "ivfpq", "ivfpqr"]
```
Comment (stephantul) [inline tests/conftest.py:11]: probably not a constant

File: tests/conftest.py [hunk 2]
```diff
@@ -8,32 +8,58 @@
...
+@pytest.fixture(params=BACKEND_PARAMS)
 def backend_type(request: pytest.FixtureRequest) -> Backend:
     """Fixture parametrizing over all backend types defined in Backend."""
     return request.param
 
 
-@pytest.fixture
-def vicinity_instance(backend_type: Backend, items: list[str], vectors: np.ndarray) -> Vicinity:
-    """Fixture creating a Vicinity instance with the given backend, items, and vectors."""
+@pytest.fixture(params=BACKEND_PARAMS, ids=BACKEND_IDS)
```
Comment (stephantul) [inline tests/conftest.py:49]: nice fixture

File: tests/test_vicinity.py [hunk 3]
```diff
@@ -9,34 +9,38 @@
 from vicinity.datatypes import Backend
 
 
-def test_vicinity_init(backend_type: Backend, items: list[str], vectors: np.ndarray) -> None:
+def test_vicinity_init(backend_type: tuple[Backend, str], items: list[str], vectors: np.ndarray) -> None:
     """
     Test Vicinity.init.
 
     :param backend_type: The backend type to use (BASIC, HNSW or Annoy).
```
Comment (stephantul) [inline tests/test_vicinity.py:16]: slightly outdated, the type changed.

File: tests/test_vicinity.py [hunk 4]
```diff
@@ -9,34 +9,38 @@
...
     vectors = np.random.default_rng(42).random((len(items) - 1, 5))
 
     with pytest.raises(ValueError):
-        vicinity = Vicinity.from_vectors_and_items(vectors, items, backend_type=backend_type)
+        vicinity = Vicinity.from_vectors_and_items(vectors, items, backend_type=backend)
 
 
-def test_vicinity_from_vectors_and_items(backend_type: Backend, items: list[str], vectors: np.ndarray) -> None:
+def test_vicinity_from_vectors_and_items(
+    backend_type: tuple[Backend, str], items: list[str], vectors: np.ndarray
```
Comment (stephantul) [inline tests/test_vicinity.py:33]: maybe define a type for this somewhere. `BackendType = tuple[Backend, str]`

File: vicinity/backends/faiss.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+    def __init__(
+        self,
+        index: faiss.Index,
+        arguments: FaissArgs,
+    ) -> None:
+        """Initialize the backend using a FAISS index."""
+        super().__init__(arguments)
+        self.index = index
+        # Enable DirectMap if specified and supported by index type
+        if isinstance(index, faiss.IndexIVF) and arguments.direct_map:
```
Comment (stephantul) [inline vicinity/backends/faiss.py:43]: so setting direct_map to `True` doesn't do anything for any index besides `IVF`. Maybe flag it or crash? 

Do you ever want to set it to False? If not, you could just not have it as an argument and enable it for IVF.

File: vicinity/backends/faiss.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+            index = faiss.IndexIVFPQR(quantizer, dim, nlist, m, nbits, m, refine_nbits)
+            index.train(vectors)
+        else:
+            raise ValueError(f"Unsupported FAISS index type: {index_type}")
+
+        index.add(vectors)
+
+        # Enable DirectMap for IVF indexes if specified
+        if isinstance(index, faiss.IndexIVF) and direct_map:
+            index.set_direct_map_type(faiss.DirectMap.Hashtable)
```
Comment (stephantul) [inline vicinity/backends/faiss.py:122]: you also call `set_direct_map_type` in the initializer, so I think the one in the initializer is redundant. Maybe?

File: vicinity/backends/faiss.py [hunk 7]
```diff
@@ -0,0 +1,229 @@
...
+    def threshold(self, vectors: npt.NDArray, threshold: float) -> list[npt.NDArray]:
+        """Query vectors within a distance threshold, using range_search if supported."""
+        out: list[npt.NDArray] = []
+
+        # Normalize query vectors if using cosine similarity
+        if self.arguments.metric == "cosine":
+            vectors = normalize(vectors)
+
+        if isinstance(
+            self.index, (faiss.IndexFlat, faiss.IndexIVFFlat, faiss.IndexScalarQuantizer, faiss.IndexIVFScalarQuantizer)
```
Comment (stephantul) [inline vicinity/backends/faiss.py:187]: maybe make this a constant somewhere.

File: vicinity/backends/faiss.py [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+                start, end = lims[i], lims[i + 1]
+                indices = I[start:end]
+                distances = D[start:end]
+
+                # Convert distances for cosine if needed
+                if self.arguments.metric == "cosine":
+                    distances = 1 - distances
+
+                # Only include indices within the threshold
+                within_threshold_indices = indices[distances < threshold]
```
Comment (stephantul) [inline vicinity/backends/faiss.py:203]: indices vs idx (line 212). Maybe rename this to idx?

File: vicinity/backends/faiss.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+        dim = vectors.shape[1]
+
+        # If using cosine, normalize vectors to unit length
+        if metric == "cosine":
+            vectors = normalize(vectors)
+            faiss_metric = faiss.METRIC_INNER_PRODUCT
+        else:
+            faiss_metric = faiss.METRIC_L2
+
+        if index_type == "flat":
```
Comment (stephantul) [inline vicinity/backends/faiss.py:83]: I think this whole tree can be simplified by 

1. recognizing that some indexes need training and others don't. Defining which ones need training and only having one `index.train` line should be sufficient. 

2. you can see if the index_type starts with `ivf`, and then remove all the boilerplate that comes with the rest.

**Review by stephantul — APPROVED** [review]
LGTM, some small changes/comments

File: vicinity/backends/faiss.py (hunk 5 above)
Comment (Pringled) [reply vicinity/backends/faiss.py:43]: Good point, let's just default it to true.

**Pull request merged**